pkg/monitor/lease.go

```diff
@@ -2,71 +2,95 @@ package monitor
 
 import (
 	"fmt"
+	"io/ioutil"
 	"math"
 	"net"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
 
-	"github.com/openshift/baremetal-runtimecfg/pkg/config"
-	"github.com/openshift/baremetal-runtimecfg/pkg/utils"
 	"github.com/sirupsen/logrus"
 	"github.com/vishvananda/netlink"
+	"gopkg.in/fsnotify.v1"
+	"gopkg.in/yaml.v2"
 )
 
 const MonitorConfFileName = "unsupported-monitor.conf"
 const leaseFile = "lease-%s"
-const macAddressBytes = 6
 
-var MacPrefixQumranet = [...]byte{0x00, 0x1A, 0x4A}
-
-type VIP struct {
-	name string
-	ip   net.IP
+type vip struct {
+	Name       string `yaml:"name"`
+	MacAddress string `yaml:"mac-address"`
+	IpAddress  string `yaml:"ip-address"`
 }
 
-func needLease(cfgPath string) (bool, error) {
+func getVipsToLease(cfgPath string) (vips *[]vip, err error) {
 	monitorConfPath := filepath.Join(filepath.Dir(cfgPath), MonitorConfFileName)
 
-	_, err := os.Stat(monitorConfPath)
+	info, err := os.Stat(monitorConfPath)
 
-	if err == nil {
+	if err == nil && info.Mode().IsRegular() {
 		log.WithFields(logrus.Fields{
 			"file": monitorConfPath,
 		}).Info("Monitor conf file exist")
-		return true, nil
+
+		data, err := ioutil.ReadFile(monitorConfPath)
+
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"filename": monitorConfPath,
+			}).WithError(err).Error("Failed to read monitor file")
+			return nil, err
+		}
+
+		return parseMonitorFile(data)
 	}
 	if os.IsNotExist(err) {
 		log.WithFields(logrus.Fields{
 			"file": monitorConfPath,
 		}).Info("Monitor conf file doesn't exist")
-		return false, nil
+		return nil, nil
 	}
 
 	log.WithFields(logrus.Fields{
 		"file": monitorConfPath,
 	}).WithError(err).Error("Failed to get file status")
-	return false, err
+	return nil, err
+}
+
+func parseMonitorFile(buffer []byte) (*[]vip, error) {
+	vips := make([]vip, 0)
+
+	if err := yaml.Unmarshal(buffer, &vips); err != nil {
+		log.WithFields(logrus.Fields{
+			"buffer": buffer,
+		}).WithError(err).Error("Failed to parse monitor file")
+		return nil, err
+	}
+
+	return &vips, nil
 }
 
-func leaseVIPs(cfgPath string, clusterName string, vips []VIP) error {
+func LeaseVIPs(log logrus.FieldLogger, cfgPath string, clusterName string, vipMasterIface string, vips []vip) error {
 	for _, vip := range vips {
-		vipFullName := GetClusterInterfaceName(clusterName, vip.name)
+		vipFullName := GetClusterInterfaceName(clusterName, vip.Name)
+
+		mac, err := net.ParseMAC(vip.MacAddress)
 
-		vipMasterIface, _, err := config.GetInterfaceAndNonVIPAddr([]net.IP{vip.ip})
 		if err != nil {
 			log.WithFields(logrus.Fields{
-				"name": vip.name,
-				"ip":   vip.ip,
-			}).WithError(err).Error("Failed to get the master device for a vip")
+				"vip": vip,
+			}).WithError(err).Error("Failed to parse mac")
```

Comment:
assuming u have 2 and one failed, what will happen? one will get lease and the other won't? If so, u have assymetry here, right? depends which on will fail first.
Not sure what the right behavior but at least I would align it - either best effort for both, or neither.

Reply:
In case anything in the leasing process would fail - the program fails as well.